pylint/checkers/len.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2016 Alexander Todorov <atodorov@MrSenko.com>
+
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+"""check for incorrect usage of len() in conditions
+"""
+import itertools
+from astroid.node_classes import Call, Const, Name, UnaryOp
+
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+
+MSGS = {
+    'W1801': ('Do not use `len(SEQUENCE)` as condition value',
+              'len-as-condition',
+              'Used when Pylint detects incorrect use of len(sequence) inside \
+              conditions.'),
+    }
+
+
+def _is_len_call(node):
+    """checks if @node is len(SOMETHING)"""
+    return isinstance(node, Call) and isinstance(node.func, Name) and node.func.name == 'len'
+
+
+class LenChecker(BaseChecker):
+    """checks for incorrect usage of len() inside conditions. pep8 states:
+    For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
+
+        Yes: if not seq:
+             if seq:
+
+        No: if len(seq):
+            if not len(seq):
+
+    Problems detected:
+    * if len(sequence):
+    * if not len(sequence):
+    * if len(sequence) == 0:
+    * if len(sequence) != 0:
+    * if len(sequence) > 0:
+    """
+
+    __implements__ = (IAstroidChecker,)
+
+    # configuration section name
+    name = 'len'
+    # messages
+    msgs = MSGS
+    priority = -2
+    # configuration options
+    options = ()
+
+    @check_messages('len-as-condition')
+    def visit_call(self, node):
+        # a len(S) call is used inside a test condition
+        # could be if, while, assert or if expression statement
+        # e.g. `if len(S):`
+        if hasattr(node.parent, 'test') and _is_len_call(node):
```

Comment:
Good catch. I've made a slight modification to detect whether or not the `len()` call is nested inside BoolOp, which at some point may have a parent with the test attribute.